staging/vhost-device-gpu/src/virt_gpu.rs

```diff
@@ -829,3 +830,68 @@ impl VirtioGpuBackend for VirtioGpu {
         Ok(OkNoData)
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::sync::{Arc, Mutex};
+
+    use super::{VirtioGpu, VirtioGpuBackend, VirtioGpuResource, VirtioGpuRing, VirtioShmRegion};
+    use rutabaga_gfx::{
+        ResourceCreateBlob, RutabagaBuilder, RutabagaComponentType, RutabagaHandler,
+    };
+    use vhost_user_backend::{VringRwLock, VringT};
+    use vm_memory::{GuestAddress, GuestMemoryAtomic, GuestMemoryMmap};
+
+    fn new_2d() -> VirtioGpu {
+        let rutabaga = RutabagaBuilder::new(RutabagaComponentType::Rutabaga2D, 0)
+            .build(RutabagaHandler::new(|_| {}), None)
+            .unwrap();
+        VirtioGpu {
+            rutabaga,
+            resources: Default::default(),
+            fence_state: Arc::new(Mutex::new(Default::default())),
+        }
+    }
+
+    #[test]
+    fn test_gpu_backend_success() {
+        let mem = GuestMemoryAtomic::new(
+            GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x1000)]).unwrap(),
+        );
+        let vring = VringRwLock::new(mem.clone(), 0x1000).unwrap();
+        vring.set_queue_info(0x100, 0x200, 0x300).unwrap();
+        vring.set_queue_ready(true);
+
+        let rutabaga1 = new_2d();
+        rutabaga1.get_capset(0, 0).unwrap();
+        //redo
+        rutabaga1.get_capset_info(0).unwrap_err();
+
+        let mut rutabaga1 = new_2d();
+        let resource_create_blob = ResourceCreateBlob::default();
+        let vecs = vec![(GuestAddress(0), 10)];
+        let mem = &GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x1000)])
+            .expect("Failed to create guest memory");
+        rutabaga1
+            .resource_create_blob(1, 1, resource_create_blob, vecs, mem)
+            .unwrap_err();
+
+        let shm_region = VirtioShmRegion::default();
+        let resource = VirtioGpuResource::default();
+        rutabaga1.resources.insert(1, resource);
+        rutabaga1.resource_map_blob(1, &shm_region, 0).unwrap_err();
+        rutabaga1.resource_unmap_blob(1, &shm_region).unwrap_err();
+        rutabaga1.context_attach_resource(1, 0).unwrap_err();
+        rutabaga1.destroy_context(1).unwrap_err();
+        rutabaga1.context_detach_resource(1, 0).unwrap_err();
+
+        rutabaga1.resource_assign_uuid(1).unwrap();
+        rutabaga1.create_context(1, 0, None).unwrap_err();
+        let mut cmd_buf = vec![0; 10];
+        let fence_ids: Vec<u64> = Vec::with_capacity(0);
+        rutabaga1
+            .submit_command(0, &mut cmd_buf[..], &fence_ids)
+            .unwrap_err();
+        rutabaga1.process_fence(VirtioGpuRing::Global, 0, 0, 0);
+    }
+}
```

Comment:
Thinking about it more. This doesn't really test a lot. I mean this calls some methods but most of them actually fail despite the test being called `_success`. The methods that are called mostly just forward to Rutabaga, so this just test that Rutabaga methods fail.  I think we could remove this test.

For example testing the `rutabaga1.process_fence` could be useful, but would require some refactoring of the code. And actually checking what process_fence does/returns. This could be tested along with the fence_handler (which would need to be refactored out the local lambda to be testable).

Reply:
i can split this test function into 2 one for success and one for failure. in calling unwrap_err it test the code that returns error, you can view it with `cargo llvm-cov test -p vhost-device-gpu -p rutabaga_gfx --open` this test covers a couple of uncovered code path, which is one of the reason we are adding test, to cover the code regions. IIRC we need to have about 70% code coverage as a requirement for a crate in rust-vmm/vhost-device.

Reply:
on second thoughts, we can remove some of the test like these ones 
```
       rutabaga1.context_attach_resource(1, 0).unwrap_err();
        rutabaga1.destroy_context(1).unwrap_err();
        rutabaga1.context_detach_resource(1, 0).unwrap_err();

        rutabaga1.resource_assign_uuid(1).unwrap();
        rutabaga1.create_context(1, 0, None).unwrap_err();
```
given that the code region has already been test by some other test in vhu_gpu.rs